### PR TITLE
Biot-Savart Cross-Foil Attention Bias: physics-derived attention structure

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Use the Biot-Savart vortex-panel induced velocities (already computed in `--vortex_panel_velocity`, PR #2357) as **additive attention biases** in the Transolver's cross-node attention. Instead of treating the induced velocity as just another input feature, inject it as a structural prior in the attention mechanism itself — nodes from fore foil and aft foil that are aerodynamically coupled get explicit attention bias toward each other.

**Why this should work:**
- PR #2357 proved that Biot-Savart information improves p_tan -3.2% and p_re -2.7% as an input feature. As an attention bias, the same physics delivers a *structural prior* on which node pairs should interact, not just additional input channels
- The attention bias approach (like ALiBi in language models) directly shapes which nodes attend to which — a Biot-Savart-derived bias tells the model "this volume node is strongly influenced by the fore foil vortex sheet" without requiring the model to infer this from features
- This is analogous to physics-informed attention in recent work (e.g., "Relative Positional Encoding via Kernel Regression", arXiv:2212.10271)
- The feature version (PR #2357) and attention bias version are **orthogonal** — they can both be active simultaneously. The feature gives local context per node; the bias shapes global attention structure.

## Instructions

You have already implemented `compute_vortex_panel_velocity()` in train.py (from PR #2357). This experiment extends it to use the inter-foil Biot-Savart magnitudes as attention biases.

### Step 1 — Add flag:
```python
biot_savart_attn_bias: bool = False   # inject Biot-Savart inter-foil coupling as attention bias
```

### Step 2 — Compute the attention bias tensor

For each batch, compute a per-pair aerodynamic coupling magnitude:
```python
def compute_biot_savart_bias(xy, is_surface, is_tandem, aoa_rad, n_panels=64):
    """Compute inter-foil attention bias from Biot-Savart kernel.
    
    Returns: [B, N, N] bias where bias[b, i, j] = induced velocity magnitude
    that foil-1 nodes contribute to foil-2 nodes and vice versa.
    For single-foil samples: all zeros.
    """
    B, N, _ = xy.shape
    bias = torch.zeros(B, N, N, device=xy.device)
    
    for b in range(B):
        if not is_tandem[b]:
            continue
        # Get surface nodes for fore foil (boundary_id == 6) and aft foil (boundary_id == 7)
        # Use is_surface mask to identify surface nodes
        # Compute vortex panels from surface nodes
        # Evaluate Biot-Savart at ALL nodes (not just surface)
        # bias[b, i, j] = |induced_velocity(from_foil_containing_j, at_node_i)|
        # For same-foil pairs: 0 (no cross-foil coupling)
        # For cross-foil pairs: Biot-Savart magnitude
        ...
    return bias  # [B, N, N]
```

The key difference from PR #2357: there, you computed the induced velocity as a node feature. Here, you compute a pairwise attention bias — the magnitude of coupling between node i (influenced) and node j (source).

**Simplified implementation:** Use the existing `compute_vortex_panel_velocity()` function as a helper. The bias for node i from foil f is the L2 norm of the induced velocity at node i from foil f's panels:
```python
# u_fore_at_all_nodes: [B, N] — induced velocity magnitude from fore foil
# For node pairs: bias[b, i, j] = u_fore[b, i] if node j is on fore foil else 0
fore_foil_nodes = (boundary_id == 6)  # [B, N]
aft_foil_nodes  = (boundary_id == 7)  # [B, N]
# Cross-foil bias: [B, N_aft, N_fore] attention bias
bias[b][aft_foil_nodes[b]][:, fore_foil_nodes[b]] = u_fore[b][aft_foil_nodes[b]].unsqueeze(-1)
```

### Step 3 — Inject into Transolver attention

In `TransolverBlock` (or `Physics_Attention_Irregular_Mesh`), accept an optional `attn_bias` argument:
```python
def forward(self, x, attn_bias=None, ...):
    # Standard attention computation
    attn_logits = Q @ K.T / sqrt(d_k)  # [B, H, N, N]
    if attn_bias is not None:
        # Add bias (broadcast across heads)
        attn_logits = attn_logits + attn_bias.unsqueeze(1)  # [B, 1, N, N]
    attn_weights = softmax(attn_logits)
    ...
```

Thread `attn_bias` from the training loop through `model.forward()` to each block.

### Step 4 — Scale the bias

The attention logits are O(1/sqrt(d_k)) scale. The Biot-Savart magnitudes (with scale=0.1 from PR #2357) may need scaling:
```python
attn_bias = biot_savart_bias * cfg.biot_savart_bias_scale  # new flag, try 1.0, 0.1, 10.0
```

Start with `--biot_savart_bias_scale 1.0`. If the model ignores the bias (attention entropy unchanged), try 10.0. If it overwhelms attention, try 0.1.

### Step 5 — Run experiments:
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/bsattn-s42" --wandb_group biot-savart-attention-bias \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --biot_savart_attn_bias --biot_savart_bias_scale 1.0

CUDA_VISIBLE_DEVICES=1 python train.py [same with --seed 73]
```

**Torch.compile note:** The `attn_bias` tensor addition is static-shape — fully compatible. The pairwise computation loop has fixed sizes.

## Baseline (PR #2357, 2-seed avg — newly merged)

| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.872 | 7.459 | 26.319 | 6.229 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

## Round 40 Bold Context

This is the **#4 bold experiment** and builds directly on your PR #2357 win. You proved Biot-Savart physics information helps (+3.2% p_tan). Now we test whether using it as an *attention structural bias* — shaping which nodes attend to which — provides additional gains on top of the input feature approach. The feature + attention bias together may compound.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Implementation Note

The PR instructions describe a full **[B, N, N]** pairwise attention bias injected into the Transolver's attention logits. However, the Transolver uses **slice attention** (O(N×S), S=96 slices) — not full N×N attention — making a [B, N, N] bias infeasible for N≈27k nodes (~7 billion entries per sample).

Instead, I implemented a lightweight approach that respects the slice-attention architecture:
- Compute Biot-Savart coupling magnitudes [B, N, 2] = (fore_foil_coupling_mag, aft_foil_coupling_mag) using the existing `compute_vortex_panel_velocity()` function
- Pass these as `"bs_feat"` in the model input dict
- In `Transolver.forward`, concatenate to `raw_xy` before each block's `spatial_bias` MLP: `raw_xy = [pos, pos, bs_mag_fore, bs_mag_aft]` → MLP → slice assignment bias
- `spatial_bias_input_dim` updated from 4 to 6 (or 8 with gap_stagger)
- Zero-init on the new weight columns ensures the BS bias starts at zero influence

This injects the Biot-Savart physics into the attention *structure* (slice routing) rather than the input features, which is the closest feasible analog to the PR's intention.

### Metrics (2-seed avg, 151 epochs, 180 min timeout)

| Metric | Baseline (PR #2357) | BS Attn Bias | Δ | % |
|--------|-------------------:|-------------:|---:|---:|
| p_in | 11.872 | 12.001 | +0.129 | +1.1% ↑ regression |
| p_oodc | 7.459 | 7.376 | -0.083 | **-1.1% ↓ improvement** |
| p_tan | 26.319 | 27.037 | +0.718 | +2.7% ↑ regression |
| p_re | 6.229 | 6.337 | +0.107 | +1.7% ↑ regression |

### Per-seed detail

**Seed 42** (W&B: `nuddpz73`):
| Split | val_loss | surf_Ux | surf_Uy | surf_p | vol_Ux | vol_Uy | vol_p |
|-------|---------|---------|---------|--------|--------|--------|-------|
| val_in_dist | 0.368 | 2.508 | 0.790 | 12.159 | 0.529 | 0.191 | 11.977 |
| val_tandem_transfer | 1.164 | 4.124 | 1.492 | 26.857 | 1.310 | 0.595 | 30.695 |
| val_ood_cond | 0.437 | 1.971 | 0.561 | 7.354 | 0.349 | 0.143 | 5.857 |
| val_ood_re | 0.416 | 1.782 | 0.412 | 6.442 | 0.520 | 0.272 | 5.502 |

**Seed 73** (W&B: `i5d2fpfr`):
| Split | val_loss | surf_Ux | surf_Uy | surf_p | vol_Ux | vol_Uy | vol_p |
|-------|---------|---------|---------|--------|--------|--------|-------|
| val_in_dist | 0.359 | 2.710 | 0.822 | 11.842 | 0.515 | 0.187 | 12.272 |
| val_tandem_transfer | 1.172 | 4.037 | 1.578 | 27.216 | 1.309 | 0.605 | 31.544 |
| val_ood_cond | 0.433 | 2.071 | 0.644 | 7.397 | 0.360 | 0.145 | 5.921 |
| val_ood_re | 0.407 | 1.942 | 0.478 | 6.231 | 0.513 | 0.268 | 5.278 |

### Command
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/bsattn-s{42,73}" --wandb_group biot-savart-attention-bias \
  --seed {42,73} --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --biot_savart_attn_bias --biot_savart_bias_scale 1.0
```

### Peak VRAM: ~41 GB (96 GB available)
### W&B run IDs: `nuddpz73` (seed 42), `i5d2fpfr` (seed 73)

## What happened

**The Biot-Savart attention bias did not improve results.** 3 of 4 metrics regressed; only p_oodc showed a marginal -1.1% improvement.

**Why it likely failed:**
1. **Redundant information**: The Biot-Savart coupling magnitudes are derived from the same `compute_vortex_panel_velocity()` used by `--vortex_panel_velocity`, which already injects 4 channels of induced velocity directly into the input features. The spatial bias MLP receives a less-informative summary (2-channel L2 norms) of the same physics already present in the main feature vector.

2. **Lightweight proxy vs. true attention bias**: The actual implementation extends the `spatial_bias` MLP input (a cheap per-node → per-slice mapping) rather than injecting a true pairwise [N, N] bias into attention logits. The slice attention architecture (O(N×S)) doesn't have a natural injection point for [N, N] biases. This proxy may be too indirect to provide structural benefit.

3. **Zero-sum competition**: Adding 2 more channels to the spatial_bias MLP (4→6) may create noise during early training that slightly hurts convergence, especially for p_tan where the model needs to route attention between both foils carefully.

## Suggested follow-ups

1. **Drop this experiment**: The attention-bias approach doesn't work in the Transolver's slice-attention architecture. The O(N×S) structure doesn't have a natural point for pairwise biases.
2. **Stick with the feature-based approach (PR #2357)**: The `--vortex_panel_velocity` feature already captures this physics effectively as an input feature.
3. **If revisiting attention structure**: Consider modifying the slice *assignment* itself to be physics-aware (e.g., a "tandem slice" that groups fore/aft foil surface nodes together) rather than biasing existing slice attention.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2367 because ADVISOR: Closing — does not beat baseline.

**W&B-verified 2-seed average:**
| Metric | 2-seed avg | Baseline | Δ |
|--------|------------|----------|---|
| p_in | 12.001 | 11.872 | +1.1% ❌ |
| p_oodc | **7.376** | 7.459 | **-1.1% ✅** |
| p_tan | 27.037 | 26.319 | +2.7% ❌ |
| p_re | 6.337 | 6.229 | +1.7% ❌ |

Your analysis is exactly right. Three factors:
1. **Redundant with existing features:** `--vortex_panel_velocity` already injects the full Biot-Savart induced velocity (4 channels) as input features. The spatial bias MLP receives a less-informative 2-channel L2 norm summary of the same physics.
2. **Architecture mismatch:** The slice-attention O(N×S) structure has no natural injection point for pairwise [N,N] biases. The proxy (extending spatial_bias MLP) is too indirect.
3. **p_tan regression (+2.7%) confirms** the attention routing was actually disrupted rather than improved by the additional channels.

Great implementation and clean analysis — the adaptation from N×N bias to spatial_bias MLP was the right engineering call. The conclusion is clear: Biot-Savart physics is best delivered as input features (PR #2357, merged), not as attention structure. This avenue is exhausted.

New bold assignment incoming..
